daemon/mgr/container.go

```diff
@@ -855,7 +863,126 @@ func (mgr *ContainerManager) Update(ctx context.Context, name string, config *ty
 
 // Upgrade upgrades a container with new image and args.
 func (mgr *ContainerManager) Upgrade(ctx context.Context, name string, config *types.ContainerUpgradeConfig) error {
-	// TODO
+	c, err := mgr.container(name)
+	if err != nil {
+		return err
+	}
+
+	c.Lock()
+	defer c.Unlock()
+
+	// check the image existed or not, and convert image id to image ref
+	image, err := mgr.ImageMgr.GetImage(ctx, config.Image)
+	if err != nil {
+		return errors.Wrap(err, "get image failed")
+	}
+
+	// FIXME: image.Name does not exist,so convert Repotags or RepoDigests to ref
+	ref := ""
+	if len(image.RepoTags) > 0 {
+		ref = image.RepoTags[0]
+	} else {
+		ref = image.RepoDigests[0]
+	}
+	config.Image = ref
+
+	var (
+		needRollback        = false
+		backupContainerMeta = *c.meta
+	)
+
+	defer func() {
+		if needRollback {
+			c.meta = &backupContainerMeta
+			if err := mgr.Client.CreateSnapshot(ctx, c.ID(), c.meta.Image); err != nil {
+				logrus.Errorf("Upgrade rollback failed to create snapshot: %v", err)
+				return
+			}
+			// FIXME: create new containerd container may failed
+			_ = mgr.createContainerdContainer(ctx, c)
+		}
+	}()
+
+	// FIXME: mergo.Merge() use AppendSlice to merge slice.
+	// that is to say, t1 = ["a", "b"], t2 = ["a", "c"], the merge
+	// result will be ["a", "b", "a", "c"]
+	// This may occur errors, just take notes to record this.
+	if err := mergo.MergeWithOverwrite(c.meta.Config, config.ContainerConfig); err != nil {
+		return errors.Wrapf(err, "failed to merge ContainerConfig")
+	}
+	if err := mergo.MergeWithOverwrite(c.meta.HostConfig, config.HostConfig); err != nil {
+		return errors.Wrapf(err, "failed to merge HostConfig")
+	}
+	c.meta.Image = config.Image
+
+	// If container is running,  we need change
+	// configuration and recreate it. Else we just store new meta
+	// into disk, next time when starts container, the new configurations
+	// will take effect.
+	if c.IsRunning() {
+		// Inherit volume configurations from old container,
+		// New volume configurations may cover the old one.
+		// c.meta.VolumesFrom = []string{c.ID()}
+
+		// FIXME: here will forcely stop container afer 3s.
+		// If DestroyContainer failed, we think the old container
+		// not changed, so just return error, no need recover it.
+		_, err := mgr.Client.DestroyContainer(ctx, c.ID(), 3)
+		if err != nil {
+			return errors.Wrapf(err, "failed to destroy container")
+		}
+
+		// remove snapshot of old container
+		if err := mgr.Client.RemoveSnapshot(ctx, c.ID()); err != nil {
+			return errors.Wrap(err, "failed to remove snapshot")
+		}
+
+		// wait util old snapshot to be deleted
+		wait := make(chan struct{})
+		go func() {
+			for {
+				// FIXME(ziren) Ensure the removed snapshot be removed
+				// by garbage collection.
+				time.Sleep(100 * time.Millisecond)
+
+				msg, err := mgr.Client.GetSnapshot(ctx, c.ID())
+				logrus.Info("snapshots %+v", msg)
```

Comment:
If err != nil, just double check whether msg has data in itself.

Reply:
if error occured , msg will be a empty Snapshot.Info{}